pkg/kubelet/kubelet_pods.go

```diff
@@ -1099,46 +1117,170 @@ func (kl *Kubelet) HandlePodCleanups() error {
+func (pk *podKillerWithSlice) IsPodPendingTerminationByPodName(podFullname string) bool {
+	pk.podKillingLock.Lock()
+	defer pk.podKillingLock.Unlock()
+	for _, pair := range pk.podsToKill {
+		if kubecontainer.GetPodFullName(pair.APIPod) == podFullname {
+			return true
+		}
+	}
+	for _, name := range pk.podTerminationMap {
+		if name == podFullname {
+			return true
+		}
+	}
+	return false
+}
+
+func (pk *podKillerWithSlice) markPodTerminated(uid string) bool {
+	pk.podKillingLock.Lock()
+	_, ok := pk.podTerminationMap[uid]
+	if ok {
+		klog.V(4).Infof("marking pod termination %q", uid)
+		delete(pk.podTerminationMap, uid)
+	}
+	pk.podKillingLock.Unlock()
+	return ok
+}
+
+// KillPod sends pod killing request to the killer
+func (pk *podKillerWithSlice) KillPod(pair *kubecontainer.PodPair) {
+	if pair.APIPod == nil {
+		pair.APIPod = pair.RunningPod.ToAPIPod()
+	}
+	pk.podKillingLock.Lock()
+	pk.podsToKill = append(pk.podsToKill, pair)
+	pk.podKillingLock.Unlock()
+	pk.cond.Signal()
+}
+
+// isPodBeingKilled adds the pod UID to set of pod UIDs which are being killed
+// It returns true when the pod UID is already in the set upon entry to the func
+func (pk *podKillerWithSlice) isPodBeingKilled(uid types.UID, podFullname string) bool {
+	pk.podKillingLock.Lock()
+	_, exists := pk.podTerminationMap[string(uid)]
+	if !exists {
+		pk.podTerminationMap[string(uid)] = podFullname
+	}
+	pk.podKillingLock.Unlock()
+	return exists
+}
+
+// wait waits for the arrival of subsequent pod killing request
+func (pk *podKillerWithSlice) wait() {
+	pk.podKillingLock.Lock()
+	defer pk.podKillingLock.Unlock()
+	for len(pk.podsToKill) == 0 {
+		pk.cond.Wait()
+	}
+}
+
+// PerformPodKillingWork launches a goroutine to kill a pod received from the slice if
 // another goroutine isn't already in action.
-func (kl *Kubelet) podKiller() {
-	killing := sets.NewString()
-	// guard for the killing set
-	lock := sync.Mutex{}
-	for podPair := range kl.podKillingCh {
-		runningPod := podPair.RunningPod
-		apiPod := podPair.APIPod
-
-		lock.Lock()
-		exists := killing.Has(string(runningPod.ID))
-		if !exists {
-			killing.Insert(string(runningPod.ID))
-		}
-		lock.Unlock()
-
-		if !exists {
-			go func(apiPod *v1.Pod, runningPod *kubecontainer.Pod) {
-				klog.V(2).Infof("Killing unwanted pod %q", runningPod.Name)
-				err := kl.killPod(apiPod, runningPod, nil, nil)
-				if err != nil {
-					klog.Errorf("Failed killing the pod %q: %v", runningPod.Name, err)
-				}
-				lock.Lock()
-				killing.Delete(string(runningPod.ID))
-				lock.Unlock()
-			}(apiPod, runningPod)
+func (pk *podKillerWithSlice) PerformPodKillingWork() {
+	var podsToKill []*kubecontainer.PodPair
+	for true {
+		pk.podKillingLock.Lock()
+		for _, podPair := range pk.podsToKill {
+			podsToKill = append(podsToKill, podPair)
+		}
+		pk.podsToKill = pk.podsToKill[:0]
+		pk.podKillingLock.Unlock()
+
+		if len(podsToKill) == 0 {
```

Comment:
we are doubling up on the locking and condition check here.  remove the `wait()` function, inline it here, and move the Unlock on 1259 below this block
```
for len(pk.podsToKill) == 0 {
		pk.cond.Wait()
}
```

Reply:
I was thinking on improving locking for this part of code.

This would be updated in the next push.